loopy/preprocess.py

```diff
@@ -1015,10 +1016,16 @@ def map_reduction_seq(expr, rec, nresults, arg_dtypes,
 
         init_insn_depends_on = frozenset()
 
-        global_barrier = lp.find_most_recent_global_barrier(temp_kernel, insn.id)
+        # check first that the original kernel had global barriers
+        # if not, we don't need to check. Since the function
+        # kernel_has_global_barriers is cached, we don't do
+        # extra work compared to not checking.
+        if kernel_has_global_barriers(kernel):
```

Comment:
```suggestion
        # FIXME: Explain why we care about global barriers her
        if kernel_has_global_barriers(kernel):
```
TBH, it is not at all clear to me why this is here at all. It may be related to efficiency (so as to avoid saving and reloading accumulators across global barriers), but it should state that. And even if it is that, I think it's kind of questionable to force that here.